File: day3/first/main.py
def isvalid(c):
    if (not c.isnumeric()) and c != '.':
        return True
    return False

def ispart(i, l, r, mat):
    if l > 0 and mat[i][l-1] != '.':
        return True
    
    if r < len(mat) - 1 and mat[i][r+1] != '.':
        return True
    
    for j in range(l-1, r+2):
        if j < 0 or j >= len(mat[0]):
            continue
        if i > 0 and isvalid(mat[i-1][j]):
            return True
        if i < len(mat) - 1 and isvalid(mat[i+1][j]):
            return True
    
    return False

def main():
    matrix = []

    with open("../input.txt") as file:
        while (line := file.readline()):
            matrix.append(line.strip())

    n, m = len(matrix), len(matrix[0])
    sum = 0

    for i in range(n):
        r = 0
        while r < m:
            if not matrix[i][r].isnumeric():
                r += 1
                continue
            l = r
            while r < m and matrix[i][r].isnumeric():
                r += 1
            if ispart(i, l, r-1, matrix):
                sum += int(matrix[i][l:r])
            r += 1

    print(sum)

main()

File: day3/first/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	input, err := os.Open("../input.txt")
	if err != nil {
		os.Exit(1)
	}
	scanner := bufio.NewScanner(input)

	matrix := [][]byte{}
	for scanner.Scan() {
		matrix = append(matrix, append([]byte{}, scanner.Text()...))
	}

	n, m := len(matrix), len(matrix[0])
	sum := 0
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if !isNumeric(matrix[i][j]) {
				continue
			}
			lj := j
			for j < m && isNumeric(matrix[i][j]) {
				j++
			}
			if isPart(i, lj, j-1, matrix) {
				partNum, _ := strconv.Atoi(string(matrix[i][lj:j]))
				sum += partNum
			}
		}
	}
	fmt.Println(sum)
}

func isNumeric(b byte) bool {
	return b >= '0' && b <= '9'
}

func isPart(i, lj, rj int, mat [][]byte) bool {
	if lj > 0 && mat[i][lj-1] != '.' {
		return true
	}
	if rj < len(mat)-1 && mat[i][rj+1] != '.' {
		return true
	}
	for j := lj - 1; j < rj+2; j++ {
		if j < 0 || j >= len(mat[0]) {
			continue
		}
		if i > 0 {
			if b := mat[i-1][j]; b != '.' && !isNumeric(b) {
				return true
			}
		}
		if i < len(mat)-1 {
			if b := mat[i+1][j]; b != '.' && !isNumeric(b) {
				return true
			}
		}
	}
	return false
}
